h2/src/main/org/h2/engine/Constants.java

```diff
@@ -267,8 +267,11 @@
     /**
      * The maximum allowed length for character string, binary string, and other
      * data types based on them; excluding LOB data types.
+     * <p>
+     * This needs to be less than (2^31-8)/2 to avoid running into the limit on
+     * encoding data fields when storing rows.
      */
-    public static final int MAX_STRING_LENGTH = 1024 * 1024;
+    public static final int MAX_STRING_LENGTH = 1024 * 1024 * 1024;
```

Comment:
It should be `1_000_000_000` as specified in documentation.

1024^3 is actually larger than actual limit of `java.lang.String` since Java 9 (for strings with at least one character with code larger than 255).